### ISSUE TITLE
Tensorflow 2.0 for raspberry pi installation

### ISSUE THREAD

**Issue by Risslock** [tensorflow/tensorflow#29704]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version:
- Python version:
- Installed using virtualenv? pip? conda?:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the problem**

**Provide the exact sequence of commands / steps that you executed before running into the problem**

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by ThomasKlein18** [comment]
I have the same issue when trying to install TensorFlow 2.0.0-rc0 via pip3, Raspberry Pi 3 B+ with raspbian 10 (buster):

Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting tensorflow==2.0.0-rc0
  Could not find a version that satisfies the requirement tensorflow==2.0.0-rc0 (from versions: 0.11.0, 1.12.0, 1.13.1)
No matching distribution found for tensorflow==2.0.0-rc0

**Comment by ll01** [comment]
> I've tried this method:
> 
> https://www.tensorflow.org/install/source_rpi
> 
> ```
> git checkout r2.0
> ```
> 
> Using python3 for the target.
> 
> But it fails:
> 
> ```
> Successfully built 798896a6c5b6
> Successfully tagged tf_ci.pi-python3:latest
> Running 'tensorflow/tools/ci_build/pi/build_raspberry_pi.sh' inside tf_ci.pi-python3...
> Reading package lists...
> Building dependency tree...
> Reading state information...
> sudo is already the newest version (1.8.16-0ubuntu1.7).
> 0 upgraded, 0 newly installed, 0 to remove and 9 not upgraded.
> addgroup: Please enter a username matching the regular expression configured
> via the NAME_REGEX[_SYSTEM] configuration variable.  Use the `--force-badname'
> option to relax this check or reconfigure NAME_REGEX.
> ```
> 
> macOS 10.14.6
> docker desktop 2.1.0.2

I get similar results as well on mac os 
`git checkout v2.0.0-rc2`
Using the python3 target outputs 

![Screenshot 2019-10-02 at 10 58 04](https://user-images.githubusercontent.com/6953881/66035719-9476ff80-e503-11e9-85ca-402c62d65a59.png)

On windows it dosn't even get to that point (I'm on my mac right now. I will post the windows results later).

**Comment by ll01** [comment]
A bit more digging and I found this guy 
https://github.com/PINTO0309/Tensorflow-bin
that was able to compile the arm version of TensorFlow 2 but I haven't checked it out yet.

**Comment by PINTO0309** [comment]
I have never been able to generate a healthy binary with the official procedure. So I always do native builds with RaspberryPi. RaspberryPi4 (Buster) can be built in about 12 hours. glibc2.29 and openjdk-8-jdk are required to use my Tensorflow binary, so please read the usage section carefully.
**https://github.com/PINTO0309/Tensorflow-bin#usage**
Please note that the wheel file I created is unofficial.
The build parameters are as follows.
**https://github.com/PINTO0309/Tensorflow-bin#build-parameter**

**Comment by gkvoelkl** [comment]
Is there any installation of TensorFlow 2.0 for Raspberry Pi?

Building does not work.

When will there be a pip wheel?

**Comment by PINTO0309** [comment]
@gkvoelkl 
**https://github.com/PINTO0309/Tensorflow-bin/raw/master/tensorflow-2.0.0-cp37-cp37m-linux_armv7l.whl**

**Comment by dougparkarosa** [comment]
Is there an estimate to when this might be fixed? I'm at a point where I'm needing to make a use/don't use TensorFlow decision with respect to the Raspberry Pi 4. The oldness of this report, combined with the incorrect official documentation that indicates 2.0 is supported through "pip install tensorflow", leaves me in a tight spot with regard to moving forward with my project.

**Comment by FlorinAndrei** [comment]
@dougparkarosa The PINTO0309 repo seems to have up to date packages. But, of course, it's not the official distribution.

https://github.com/PINTO0309/Tensorflow-bin/

**Comment by PINTO0309** [comment]
@FlorinAndrei 
Since the official wheel will not be released indefinitely, it was generated on its own. I also created it to speed up Tensorflow Lite. If you need to use the official one, you have to wait a lot.

**Comment by dougparkarosa** [comment]
Tried PINTO0309 package on my Pi 4. Ran the Fashion tutorial and it seemed to work. Can't say how solid it is at this point, but looks promising. And probably enough to keep me moving forward, for now.

**Comment by dougparkarosa** [comment]
Tried PINTO0309 package On my pi B+. Same success there.

**Comment by ThomasKlein18** [comment]
PINTO0309 package worked for me as well, thank you! Pi B+ Raspbian 10 Buster. At first, when importing TensorFlow, I saw the error 
"libf77blas.so.3: cannot open shared object file: No such file or directory"
but `sudo apt-get install libatlas-base-dev` solved the problem - that's probably unrelated to the issue at hand, just mentioning this in case somebody else has the same problem.

**Comment by PINTO0309** [comment]
I'll let you know just in case. My repository README already mentions installing libatlas-base-dev for usage.
**https://github.com/PINTO0309/Tensorflow-bin/#usage**

**Comment by ThomasKlein18** [comment]
Oh thanks! I overlooked that, I just pip-installed the .whl file you linked to above, which actually caused a problem because I hadn't installed h5py. But by reading your README I was able to resolve that problem! Thanks again!

**Comment by aeropia** [comment]
@PINTO0309 
I've tried in the usage you had suggested but found errors in the final step installing .whl file. It looks I have problem building scipy package. Anyone knows how to fix the problem?

Building wheels for collected packages: scipy
  Building wheel for scipy (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/bin/python3.7 /usr/local/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmpwtronura
       cwd: /tmp/pip-install-r2qva1ao/scipy
  Complete output (4511 lines):
  lapack_opt_info:
  lapack_mkl_info:
  customize UnixCCompiler
    libraries mkl_rt not found in ['/usr/local/lib', '/usr/lib', '/usr/lib/arm-linux-gnueabihf']
    NOT AVAILABLE

  openblas_lapack_info:
  customize UnixCCompiler
  customize UnixCCompiler
    libraries openblas not found in ['/usr/local/lib', '/usr/lib', '/usr/lib/arm-linux-gnueabihf']
    NOT AVAILABLE

  openblas_clapack_info:
  customize UnixCCompiler
  customize UnixCCompiler
    libraries openblas,lapack not found in ['/usr/local/lib', '/usr/lib', '/usr/lib/arm-linux-gnueabihf']
    NOT AVAILABLE

**Comment by PINTO0309** [comment]
@aeropia
Please refer to the following issue. Solves the same problem as you.
**[Can't install 2.1.0 on Python 3.7 Raspberry Pi 4 #19](https://github.com/PINTO0309/Tensorflow-bin/issues/19)**

**Comment by aeropia** [comment]
@PINTO0309 
It looks the issue has been resolved and now being installed. Thanks!

**Comment by aeropia** [comment]
@PINTO0309 
I have another error message : ERROR: tensorboard 2.1.0 has requirement setuptools>=41.0.0, but you'll have setuptools 40.6.2 which is incomp. 
Is this an error expected?

**Comment by PINTO0309** [comment]
Try the following:
```
$ sudo pip3 install setuptools --upgrade
```

**https://pypi.org/project/setuptools/**

**Comment by aeropia** [comment]
@PINTO0309 
Got it. Thanks again!

**Comment by flo-ks** [comment]
Still not fixed!
when I run `pip3 install tensorflow` it installs v1.14.0!
When I try to run `pip3 install tensorflow==2.1.0` I get the described error.

System:
Raspbian 10 buster,
python version: 3.7.3,
pip version: 20.0.2

**Comment by jingw222** [comment]
Plus, the installation guide on official tensorflow website for the version 2 says it supports 64-bit system of raspbian 9+. What's that mean? I don't think raspbian as of now is 64-bit yet.

**Comment by dynamicwebpaige** [comment]
Are you attempting to train TF 2.x models on a Raspberry Pi, or just execute them? If the latter, [TF Lite supports TF 2.x SavedModels](https://www.tensorflow.org/lite/guide/build_rpi).

**Comment by Saduf2019** [comment]
@Risslock 
Is this still an issue.

**Comment by KopfKrieg** [comment]
Tensorflow 2.x is still not available as pre-built package for the Raspberry Pi. I'd say it's still an issue.

**Comment by Risslock** [comment]
> Tensorflow 2.x is still not available as pre-built package for the Raspberry Pi. I'd say it's still an issue.



> @Risslock
> Is this still an issue.

As KopfKrieg said. Even though you can compile it, it is still not available as pre-built package for the Raspberry Pi

**Comment by Python-Ninja-Hebi** [comment]
When will it be available?

**Comment by KopfKrieg** [comment]
Update: There *are* pre-built packages for the Raspberry Pi available, the urls (not even links you can just click!) are at the bottom of this page: https://www.tensorflow.org/install/pip

Those packages aren't available via pip/piwheels, though (and that's the real issue for me)

**Comment by oholimoli** [comment]
Why is no official `.whl` package for the Raspberry Pi 4 available? => https://www.tensorflow.org/install/pip

**Comment by KopfKrieg** [comment]
Update to my previous comment: Building the Tensorflow 2.x package is not possible as described in the official docs, it will only result in an error. Installing the provided package is also not possible on a current Raspbian 10 OS because it's not compatible with Python 3.7.

Tl;dr: Still no official way of installing Tensorflow 2.x.

**Comment by TylerADavis** [comment]
@KopfKrieg What error are you seeing when trying to build? I'm experiencing an error when trying to build as well, and I'm wondering if it's the same one.

I'm getting  the following (also described in issue #44028)
```
addgroup: Please enter a username matching the regular expression configured
via the NAME_REGEX[_SYSTEM] configuration variable.  Use the `--force-badname'
option to relax this check or reconfigure NAME_REGEX.
```

**Comment by KopfKrieg** [comment]
> What error are you seeing when trying to build? I'm experiencing an error when trying to build as well, and I'm wondering if it's the same one.

No idea, sorry. I currently don't have time to look further into that issue, all I know is building tensorflow 2.x didn't work for me.

**Comment by KopfKrieg** [comment]
I was able to build Tensorflow 2.4 for Python 3.5 for the Raspberry Pi 3 using the docker method (and it only took 8 hours...). Unfortunately, Tensorflow 2.4 requires Numpy 1.19.2 or higher, but Numpy dropped Python 3.5 support in the 1.19.x release.

Tl;dr: I think this issue should be closed, because even if you can build Tensorflow, you still can't build the required dependencies. Also, Python 3.5 is now unsupported.

**Comment by ddelange** [comment]
For future reference, just successfully built a py37/tf2 docker image for `linux/arm/v7` based on `debian:buster-slim`.

This setup does not need to build any package from source:

```Dockerfile
# install python3.7 and expose latest pip
RUN apt-get update && apt-get install -y --no-install-recommends \
    python3-pip && \
    pip3 install -U --no-cache-dir pip
    # cleanup
    rm -rf /var/lib/apt/lists/* && \

# avoid source distributions as much as possible by looking for wheels at piwheels.org
ENV PIP_EXTRA_INDEX_URL=https://piwheels.org/simple

# neither pypi.org/project/tensorflow nor piwheels.org/project/tensorflow has tf2 wheels for armv7l
# but luckily piwheels.org has wheels for numpy, scipy etc and github.com/lhelontra/tensorflow-on-arm has a tf2 wheel
RUN pip install --no-cache-dir https://github.com/lhelontra/tensorflow-on-arm/releases/download/v2.4.0/tensorflow-2.4.0-cp37-none-linux_armv7l.whl
```

**Comment by COfek** [comment]
> ### tldr: get your tf2 wheel from https://github.com/lhelontra/tensorflow-on-arm/releases
> For future reference, this setup does not need to build any package from source, by pointing pip to arm wheels:
> 
> ```dockerfile
> # no need to add the 'arm32v7/' prefix when using something like docker buildx --platform="linux/arm/v7"
> # because the plain debian tags are multi-arch https://hub.docker.com/_/debian?tab=tags
> FROM arm32v7/debian:buster-slim
> 
> # install python3.7 and expose latest pip
> RUN apt-get update && \
>     apt-get install -y --no-install-recommends python3-pip && \
>     pip3 install -U --no-cache-dir pip && \
>     # cleanup
>     rm -rf /var/lib/apt/lists/*
> 
> # avoid source distributions as much as possible by looking for wheels at piwheels.org
> ENV PIP_EXTRA_INDEX_URL=https://piwheels.org/simple
> 
> # neither pypi.org/project/tensorflow nor piwheels.org/project/tensorflow has tf2 wheels for armv7l
> # but luckily piwheels.org has wheels for numpy, scipy etc and github.com/lhelontra/tensorflow-on-arm has a tf2 wheel
> RUN pip install --no-cache-dir https://github.com/lhelontra/tensorflow-on-arm/releases/download/v2.4.0/tensorflow-2.4.0-cp37-none-linux_armv7l.whl
> ```

Hello, I've had the same problem happen to me and I've seen your comment.
I'm not very experienced with managing packages and using wheels so I am saying sorry in advanced for the noob question.
what does the first line mean?
`FROM arm32v7/debian:buster-slim` isn't a line I can execute in the terminal 
the only line which is executable for me is `pip install --no-cache-dir https://github.com/lhelontra/tensorflow-on-arm/releases/download/v2.4.0/tensorflow-2.4.0-cp37-none-linux_armv7l.whl`

my on the raspberry python3.7.3 is installed and pip is updated to 21.0.1 so is the second phase unnecessary because it doesn't seem to run for me as well.

Thanks in advance for the response.

**Comment by ddelange** [comment]
Hi @COfek, the above snippet is a Dockerfile, so not intended to be interpreted by a shell. For you indeed only the last RUN statement is relevant. The ENV statement with piwheels.org is set by default on Raspbian, so only needed within docker: you can ignore it. Maybe you're running an ARM v6 kernel. Try changing the wheel link from v7 to v6, maybe that helps.

**Comment by COfek** [comment]
@ddelange uname -r returns 5.10.14-v7+

**Comment by KopfKrieg** [comment]
@ddelange The problem is those are not official packages. I prefer/require packages that are either available from piwheels.org or from tensorflow.org, but not from a random repository. I could also build the package myself (as explained in a previous comment, using the docker method), but it'd be way easier if the tensorflow team simply provides official whl-files.

**Comment by mihaimaruseac** [comment]
Providing official wheels means someone has to do maintenance for those. We currently don't have headcount for this position.

**Comment by yoyojacky** [comment]
I have the same issue too : 
Raspberry Pi 4B, GCC version: gcc version 10.2.1 20210110 (Debian 10.2.1-6)
Kernel: Linux raspberrypi 5.15.76-v8+ #1597 SMP PREEMPT Fri Nov 4 12:16:41 GMT 2022 aarch64 GNU/Linux
Release of OS:  Debian GNU/Linux 11 (bullseye)
Python version:   Python 3.9.2
pip 22.3 from /home/pi/.local/lib/python3.9/site-packages/pip (python 3.9)

I can not install tensorflow==2.2.0 to my system, but i am working on building my donkeycar, it says we need install tensorflow==2.2.0 , but it can not be due to the python version has been update to 3.9.2, but tensorflow2.2.0 needs python3.7,  so how can i install tensorflow==2.2.0 to my python3.9 environment ?

**Comment by ddelange** [comment]
tensordlow 2.2.0 was released in may 2020, that's pretty old. that version has wheels [available](https://pypi.org/project/tensorflow/2.2.0/#files) for py 3.5 until 3.8, but not for aarch64 (raspberry pi). so you would have to build from source plus it was untested/unsupported at that point in time. you (or your upstream dependency) would need to switch to a more recent tensorflow version (>=2.10.0), which have pre-built and tested aarch64 wheels available for py 3.7 until 3.10.
